File: scholarzim/src/main/java/com/scholarzim/controller/ReportController.java
package com.scholarzim.controller;

import com.scholarzim.service.ExcelReportService;
import com.scholarzim.service.ReportService;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;

@Controller
public class ReportController {

    private static final MediaType XLSX = MediaType.parseMediaType(
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");

    private final ReportService reportService;
    private final ExcelReportService excelReportService;

    public ReportController(
            ReportService reportService,
            ExcelReportService excelReportService) {

        this.reportService = reportService;
        this.excelReportService = excelReportService;
    }

    @GetMapping("/admin/reports/users.pdf")
    public ResponseEntity<byte[]> usersReport() {
        return pdf(reportService.usersReportPdf(), "users-report.pdf");
    }

    @GetMapping("/admin/reports/opportunities.pdf")
    public ResponseEntity<byte[]> opportunitiesReport() {
        return pdf(reportService.opportunitiesReportPdf(), "opportunities-report.pdf");
    }

    @GetMapping("/admin/reports/applications.pdf")
    public ResponseEntity<byte[]> applicationsReport() {
        return pdf(reportService.applicationsReportPdf(), "applications-report.pdf");
    }

    @GetMapping("/admin/reports/recommendations.pdf")
    public ResponseEntity<byte[]> recommendationsReport() {
        return pdf(reportService.recommendationsReportPdf(), "recommendations-report.pdf");
    }

    @GetMapping("/admin/reports/users.xlsx")
    public ResponseEntity<byte[]> usersExcel() {
        return excel(excelReportService.usersExcel(), "users-report.xlsx");
    }

    @GetMapping("/admin/reports/opportunities.xlsx")
    public ResponseEntity<byte[]> opportunitiesExcel() {
        return excel(excelReportService.opportunitiesExcel(), "opportunities-report.xlsx");
    }

    @GetMapping("/admin/reports/applications.xlsx")
    public ResponseEntity<byte[]> applicationsExcel() {
        return excel(excelReportService.applicationsExcel(), "applications-report.xlsx");
    }

    private ResponseEntity<byte[]> pdf(byte[] body, String filename) {
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION,
                        "attachment; filename=\"" + filename + "\"")
                .contentType(MediaType.APPLICATION_PDF)
                .body(body);
    }

    private ResponseEntity<byte[]> excel(byte[] body, String filename) {
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION,
                        "attachment; filename=\"" + filename + "\"")
                .contentType(XLSX)
                .body(body);
    }
}

File: app/Http/Controllers/Admin/ReportController.php
<?php

namespace App\Http\Controllers\Admin;

use App\Http\Controllers\Controller;
use App\Services\ExcelReportService;
use App\Services\ReportService;
use Illuminate\Support\Facades\Gate;
use Symfony\Component\HttpFoundation\Response;

class ReportController extends Controller
{
    private const XLSX = 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet';

    public function __construct(
        private readonly ReportService $reportService,
        private readonly ExcelReportService $excelReportService,
    ) {
    }

    public function hub()
    {
        return view('admin.reports');
    }

    /**
     * Every export on this controller crosses the whole platform, so the ability
     * is asserted here as well as on the admin route group. The gate is the
     * statement of the rule; the route is where it currently happens to be
     * enforced, and those are not the same thing.
     */
    private function authorizeExport(): void
    {
        abort_unless(Gate::allows('export-reports'), 403);
    }

    public function usersPdf(): Response
    {
        return $this->download($this->reportService->usersReportPdf(), 'users-report.pdf', 'application/pdf');
    }

    public function opportunitiesPdf(): Response
    {
        return $this->download($this->reportService->opportunitiesReportPdf(), 'opportunities-report.pdf', 'application/pdf');
    }

    public function applicationsPdf(): Response
    {
        return $this->download($this->reportService->applicationsReportPdf(), 'applications-report.pdf', 'application/pdf');
    }

    public function recommendationsPdf(): Response
    {
        return $this->download($this->reportService->recommendationsReportPdf(), 'recommendations-report.pdf', 'application/pdf');
    }

    public function usersExcel(): Response
    {
        return $this->download($this->excelReportService->usersExcel(), 'users-report.xlsx', self::XLSX);
    }

    public function opportunitiesExcel(): Response
    {
        return $this->download($this->excelReportService->opportunitiesExcel(), 'opportunities-report.xlsx', self::XLSX);
    }

    public function applicationsExcel(): Response
    {
        return $this->download($this->excelReportService->applicationsExcel(), 'applications-report.xlsx', self::XLSX);
    }

    private function download(string $body, string $filename, string $contentType): Response
    {
        $this->authorizeExport();

        return response($body, 200, [
            'Content-Type' => $contentType,
            'Content-Disposition' => 'attachment; filename="' . $filename . '"',
            'Content-Length' => (string) strlen($body),
        ]);
    }
}
